### ISSUE TITLE
Requests for improvements over original engine

### ISSUE THREAD

**Issue by tgjones** [OpenSAGE/OpenSAGE#380]
This issue is to track improvements over the original SAGE that have been requested. No promises to do any of them! But at least if they're all listed here we won't lose them.

* [ ] Report on misspelled command button names (and in general any asset references that can't be resolved)
* [ ] "Can you make so that the actual scaffoldings do not appear until the builder arrives (optional ofc, something you can turn on and off in the inis)? So that you can't hijack a dozer and quickly place a CC at your base and then finish it with Worker, or so that you can't block enemy building placement with scaffoldings, or so that you can't remove scrap boxes with scaffoldings? Some mods would really like to use such a change." [From YouTube](https://www.youtube.com/watch?v=h0BrXLbQRiU&lc=Ugwj4qMKF698FLx6lKJ4AaABAg)

**Comment by ruurdnijdam1986** [comment]
top down view/cara placement
bigger zoom
larger resolution

**Comment by Continous** [comment]
Un-hardcoding many graphical values would be rather nice. Shadow map resolution in example.

**Comment by BSG-75** [comment]
@Continous since OpenSage aims to support all Sage games, that shouldn't be a big problem because some stuffs aren't hardcore in later games anymore ;) For example in RA3 shadow map resolution can be modified

**Comment by Continous** [comment]
Resolution is hard coded in the games up to CnC Tiberium wars iirc too. It is just an example.

Also, an actual observer mode would be nice, with useful UI features.

**Comment by tgjones** [comment]
Can you explain to me like I’m a 5 year old what you mean by useful UI features for an observer mode? I’m not familiar at all with the state of the art in that area.

**Comment by Continous** [comment]
In most observer modes you can see;

The players' resources.
Where the players are looking.
Each players APM.

I would also like to be able to see what each player has in production, and unit counts would be pretty useful too. The ideal is basically Starcraft 2's UI as they have it for tournaments like WCS.

Shadow maps are the big one, but resolution would be another. Framerate should be un-tied from physics/gametime where ever it happens to be so tied.

**Comment by BSG-75** [comment]
@Continous didn't all these exist in the original games' observe mode, too? (Except APM)

**Comment by Continous** [comment]
Not in the ways I'm proposing, if at all. The players resources is only available if you select that specific players unit's or if you use an external mod/tool.

Further, no income over time metric is provided.

Production is only provided in RA3 and Tiberium Wars/Kane's Wrath. Generals and BFME have no such functionality. Further, it cannot show you an overview of what is in production, or what production structures/tech structures exist.

I don't think any game can show the players' screen locations on the minimap.

Unit counts are also non-existent.

Take a look at some of the WCS championships and you'll see what I mean. It's a fairly useful.

To provide and ELI5;

An overlay with all the important stuff. Player's money, tech, army count, apm, worker and resource structure count, active production, and overlay of where they're looking on the minimap.

This information is invaluable to a caster.

**Comment by Continous** [comment]
I'd like to have animated texture support added to water.

**Comment by BSG-75** [comment]
> The players resources is only available if you select that specific players unit's or if you use an external mod/tool

If you mean player's available cash, you can see it by switching to player's perspective when observing / watching replay
> Further, it cannot show you an overview of what is in production, or what production structures/tech structures exist.

I didn't play old C&C games too much, and I agree that these features should be made available for older games too, but I'm pretty sure you can see these stuffs in RA3...
> I don't think any game can show the players' screen locations on the minimap.

In RA3 it does can show player's screen location on the minimap, you just need to set the player you want to follow in the replay toolbar.

> An overlay with all the important stuff. Player's money, tech, army count, apm, worker and resource structure count, active production, and overlay of where they're looking on the minimap.
This information is invaluable to a caster.

I also think these features will be very useful, but probably we also need to find out if some of them already exists in SAGE games (and whether if they are good enough or they need improvement), so we can avoid unnecessary reinventing stuffs

**Comment by Continous** [comment]
> If you mean player's available cash, you can see it by switching to player's perspective when observing / watching replay

I specifically want something that shows an overview of _each_ player's current resources. Not just one at a time.

> I didn't play old C&C games too much, and I agree that these features should be made available for older games too, but I'm pretty sure you can see these stuffs in RA3...

I'm pretty sure active research is hard to spot, if at all available in RA3, and again is only available if you select the relevant structure.

> In RA3 it does can show player's screen location on the minimap, you just need to set the player you want to follow in the replay toolbar.

Which is not quite what I want. I'd want separate screen position icon...things in the minimap for each player plus the observer.

> I also think these features will be very useful, but probably we also need to find out if some of them already exists in SAGE games

Some may exist in the later games, but backporting would be useful.

**Comment by chemdude8** [comment]
Ability to fill units more effectively (25 rocket inf on screen with 5 humvees, should be able to select all rocketInf and issue one command to fill the 5 humvees)

More team colors

Shift+click building placement to queue multiples of the same building 
 - line-drag building placement

'advanced guard' units should:
 - 1) when presented with multiple near/in-range targets target the most effective (quads killing infantry, scorpions/rocket buggies attacking tanks)
 - 2) when using guard units should avoid overkill (think 50 buggies vs 3 tanks, currently they all shoot whatever is closest even if there's already 90 rockets heading at that unit)

Way to default upgrades for created units (i.e. when the warfactory finishes a overlord, that overlord should automatically start its gattling cannon upgrade)

Infinite/repeating build queues

Ability to send allies funds/units(?) (rule-toggleable)

Advanced multi-player options (I.e. allow specifying mods to basic maps from the lobby (removing auroras/Sw/Db/Nm))  -- Guessing the existing framework could be leveraged and simply make a .map file which gets deleted at the end of the map/start of the menus?

Support for No Rush (Make terrain (and air units) unpassable for x amount of time y distance from player start locations)

An 'unholy alliance' mode ala redAlert2 (start with more than one faction) -- Could be achieved if the 'mod-map' allowed spawning units.

Enhanced friend interface

A return to lobby after match option - this would be nice for AOD's/compstomps.

Options for 'handicaps' i.e. player 1 gets 5k starting money, player 2 gets 10k starting money, player1 unit costs 125% player2 unit costs 75%

Dual monitor support (second monitor showing mirroring mini-map or picture-in-picture frame (use the camera position save theory))

Camera position saves (alt+shift+1 to save, alt+1 to jump to that camera location)

Camera picture in picture

Support for 'sharing' units on a team (i.e. allow issuing orders to allies units)

Support for 'team general abilities' 
 -  allow for making maps/mods with a tech tree that teams can contribute to (ala demigod)
 -  could be $ based or could be exp based

Ability to message from a game lobby back to the multiplayer-channels

Display for average income rate over the past x seconds in game

Ability to buy full units/bunkers (humvees with rocket inf inside)  -- probably a 'rules toggle'

**Comment by chemdude8** [comment]
Also improvements such as sorting/filtering/searching available maps (perhaps split by number of players)

**Comment by chemdude8** [comment]
Mod manager, allow for swapping between multiple mods in-game (bonus points if you can join multi-player lobbies and have it switch to that mod for you).

**Comment by ClementJ18** [comment]
I think some sort of string templating system would greatly help game/mod developers. By string templating, I mean the ability to insert a macro into a string by referencing its name. A lot of developers struggle to keep the game strings up to date especially when changes are rapidly introduced.

An example:
gamedata.ini
```
#define ABILITY_ARMOR_BUFF 30%
```

lotr.str
```
CONTROLBAR:AbilityThatGivesArmorBuffDescription
"When activated grants +%ABILITY_ARMOR_BUFF armor"
End
```

In game it would then look like: "When activated grants +30% armor"
